include/revng/Pipeline/Step.h

```diff
@@ -0,0 +1,387 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <memory>
+#include <set>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Error.h"
+
+#include "revng/ADT/STLExtras.h"
+#include "revng/Pipeline/Containers.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/InputOutputContract.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/FunctionIntrospection.h"
+
+namespace Pipeline {
+
+template<typename T>
+class NamedPtr {
+public:
+  using value_type = T;
+  NamedPtr(T *Obj, llvm::StringRef Name) : Obj(Obj), Name(Name) {}
+
+  T *operator->() const { return Obj; }
+  T &operator*() const { return *Obj; }
+
+  llvm::StringRef getName() const { return Name; }
+
+private:
+  T *Obj;
+  llvm::StringRef Name;
+};
+
+namespace Detail {
+using StringArrayRef = llvm::ArrayRef<std::string>;
+
+template<typename T>
+auto &getContainerFromName(Containers &Containers, llvm::StringRef Name) {
+  return Containers.getOrCreate<T>(Name);
+}
+
+template<typename T>
+concept NamedPtrConcept = is_specialization_v<T, NamedPtr>;
+
+template<NamedPtrConcept T>
+auto getContainerFromName(Containers &Containers, llvm::StringRef Name) {
+  using Requested = std::decay_t<typename T::value_type>;
+  Requested &Container = Containers.getOrCreate<Requested>(Name);
+
+  return T(&Container, Name);
+}
+
+template<typename Pipe, typename... Args, size_t... S>
+void invokeImpl(Context &Ctx,
+                Pipe &Enf,
+                void (Pipe::*F)(Context &, Args...),
+                Containers &Containers,
+                const StringArrayRef &ArgsNames,
+                const std::integer_sequence<size_t, S...> &) {
+  (Enf
+   .*F)(Ctx,
+        getContainerFromName<std::decay_t<Args>>(Containers, ArgsNames[S])...);
+}
+
+template<typename Pipe, typename... Args, size_t... S>
+void invokeImpl(const Context &Ctx,
+                Pipe &Enf,
+                void (Pipe::*F)(const Context &, Args...),
+                Containers &Containers,
+                const StringArrayRef &ArgsNames,
+                const std::integer_sequence<size_t, S...> &) {
+  (Enf
+   .*F)(Ctx,
+        getContainerFromName<std::decay_t<Args>>(Containers, ArgsNames[S])...);
+}
+
+template<typename Pipe, typename... Args, size_t... S>
+void invokeImpl(Pipe &Enf,
+                void (Pipe::*F)(Args...),
+                Containers &Containers,
+                const StringArrayRef &ArgsNames,
+                const std::integer_sequence<size_t, S...> &) {
+  (Enf
+   .*F)(getContainerFromName<std::decay_t<Args>>(Containers, ArgsNames[S])...);
+}
```

Comment:
Terrible indentation

Reply:
there is like, literally no other line break acceptable to check convetions. i can reduce the size of the names, but then those become less readable

Reply:
You can declare a variable for `Enf.*F` and a `using` for `std::decay_t` or maybe a local `using namespace std`.